File: leetcode-algorithms/448. Find All Numbers Disappeared in an Arra/Solution.cpp
// Given an array of integers where 1 ≤ a[i] ≤ n (n = size of array), some elements appear twice and others appear once.
//
// Find all the elements of [1, n] inclusive that do not appear in this array.
//
// Could you do it without extra space and in O(n) runtime? You may assume the returned list does not count as extra space.
//
// Example:
//
// Input:
// [4,3,2,7,8,2,3,1]
//
// Output:
// [5,6]
class Solution {
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        for (int i = 0; i < nums.size(); i++) {
            while ((nums[i] != (i + 1)) && (nums[i] != nums[nums[i] - 1])) {
                swap(nums[i], nums[nums[i] - 1]);
            }
        }

        vector<int> rslt;

        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] != (i + 1)) rslt.push_back(i + 1);
        }

        return rslt;
    }
};

File: leetcode-algorithms/448. Find All Numbers Disappeared in an Arra/Solution.java
class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
        	while ((nums[i] != i + 1) && (nums[nums[i] - 1] != nums[i])) {
        		int v = nums[i];
        		nums[i] = nums[nums[i] - 1];
        		nums[v - 1] = v;
        	}
        }

        List<Integer> ar = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
        	if (nums[i] != i + 1) ar.add(i+1);
        }

        return ar;
    }
}
